shopinvader/components/access_info.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2019 Camptocamp (http://www.camptocamp.com).
+# @author Simone Orsi <simone.orsi@camptocamp.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.addons.component.core import Component
+
+
+class PartnerAccess(Component):
+    """Define access rules to partner from client side."""
+
+    _name = "shopinvader.partner.access"
+    _inherit = "base.shopinvader.component"
+    _usage = "access.info"
+    _apply_on = "res.partner"
+
+    @property
+    def service_work(self):
+        return getattr(self.work, "service_work", None)
+
+    @property
+    def partner(self):
+        return getattr(self.work, "partner", None)
+
+    @property
+    def partner_user(self):
+        return getattr(self.work, "partner_user", self.partner)
+
+    def is_main_partner(self):
+        return self.partner == self.partner_user
+
+    def is_owner(self, partner_id):
+        return partner_id == self.partner_user.id
+
+    def for_profile(self, partner_id):
+        info = {"read": True, "update": True, "delete": False}
+        if not self.is_main_partner() and partner_id != self.partner_user.id:
+            info["update"] = False
+        return info
+
+    def for_address(self, address_id):
+        info = {"read": True, "update": True, "delete": True}
```

Comment:
@simahawk Can we really delete an address? IMO an address can be deleted only if it's not used... (linked to a SO)

Reply:
@lmignon by default you can delete it. There's no blocker AFAIK.
If you want to add more advanced checks: fine, this is the base for them ;)

Reply:
I hope we have a constraint ;-) 
Your proposal seems good to me...

Reply:
@lmignon I tested: there are sql constraints for all kind of partners on orders (partner_id, partner_shipping_id, partner_invoice_id).

I guess we can simply add this check here:

`sale.order.search_count(['|', ('partner_shipping_id','=', address_id), ('partner_invoice_id','=', address_id)])`

Fine for you?

Reply:
hummmm.... The current implementation is not designed to allow batch processing... That means we'd have as many queries as there are addresses. It is perhaps more interesting not to make a check and let the error message pop up when deleting an address if it is explicit enough. :thinking:

Reply:
@lmignon actually they are cached and THIS is the issue because you won't get the right permission until you invalidate it.
Let's keep it as it is now, then we'll see.

Reply:
:+1: